File: src/backend/compras/src/model/produto_model.py
from src import db
from src.model.produto_marca_model import ProdutoMarcaModel
from src.model.tribut_icms_custom_cab_model import TributIcmsCustomCabModel
from src.model.tribut_grupo_tributario_model import TributGrupoTributarioModel
from src.model.produto_unidade_model import ProdutoUnidadeModel
from src.model.produto_subgrupo_model import ProdutoSubgrupoModel


class ProdutoModel(db.Model):
    __tablename__ = 'produto'

    id = db.Column(db.Integer, primary_key=True)
    nome = db.Column(db.String(100))
    descricao = db.Column(db.String(250))
    gtin = db.Column(db.String(14))
    codigo_interno = db.Column(db.String(50))
    valor_compra = db.Column(db.Float)
    valor_venda = db.Column(db.Float)
    codigo_ncm = db.Column(db.String(8))
    estoque_minimo = db.Column(db.Float)
    estoque_maximo = db.Column(db.Float)
    quantidade_estoque = db.Column(db.Float)
    data_cadastro = db.Column(db.DateTime)
    id_produto_marca = db.Column(db.Integer, db.ForeignKey('produto_marca.id'))
    id_tribut_icms_custom_cab = db.Column(db.Integer, db.ForeignKey('tribut_icms_custom_cab.id'))
    id_tribut_grupo_tributario = db.Column(db.Integer, db.ForeignKey('tribut_grupo_tributario.id'))
    id_produto_unidade = db.Column(db.Integer, db.ForeignKey('produto_unidade.id'))
    id_produto_subgrupo = db.Column(db.Integer, db.ForeignKey('produto_subgrupo.id'))

    produto_marca_model = db.relationship('ProdutoMarcaModel', foreign_keys=[id_produto_marca])
    tribut_icms_custom_cab_model = db.relationship('TributIcmsCustomCabModel', foreign_keys=[id_tribut_icms_custom_cab])
    tribut_grupo_tributario_model = db.relationship('TributGrupoTributarioModel', foreign_keys=[id_tribut_grupo_tributario])
    produto_unidade_model = db.relationship('ProdutoUnidadeModel', foreign_keys=[id_produto_unidade])
    produto_subgrupo_model = db.relationship('ProdutoSubgrupoModel', foreign_keys=[id_produto_subgrupo])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_produto_subgrupo = data.get('idProdutoSubgrupo')
        self.id_produto_marca = data.get('idProdutoMarca')
        self.id_produto_unidade = data.get('idProdutoUnidade')
        self.id_tribut_icms_custom_cab = data.get('idTributIcmsCustomCab')
        self.id_tribut_grupo_tributario = data.get('idTributGrupoTributario')
        self.nome = data.get('nome')
        self.descricao = data.get('descricao')
        self.gtin = data.get('gtin')
        self.codigo_interno = data.get('codigoInterno')
        self.valor_compra = data.get('valorCompra')
        self.valor_venda = data.get('valorVenda')
        self.codigo_ncm = data.get('codigoNcm')
        self.estoque_minimo = data.get('estoqueMinimo')
        self.estoque_maximo = data.get('estoqueMaximo')
        self.quantidade_estoque = data.get('quantidadeEstoque')
        self.data_cadastro = data.get('dataCadastro')

    def serialize(self):
        return {
            'id': self.id,
            'idProdutoSubgrupo': self.id_produto_subgrupo,
            'idProdutoMarca': self.id_produto_marca,
            'idProdutoUnidade': self.id_produto_unidade,
            'idTributIcmsCustomCab': self.id_tribut_icms_custom_cab,
            'idTributGrupoTributario': self.id_tribut_grupo_tributario,
            'nome': self.nome,
            'descricao': self.descricao,
            'gtin': self.gtin,
            'codigoInterno': self.codigo_interno,
            'valorCompra': self.valor_compra,
            'valorVenda': self.valor_venda,
            'codigoNcm': self.codigo_ncm,
            'estoqueMinimo': self.estoque_minimo,
            'estoqueMaximo': self.estoque_maximo,
            'quantidadeEstoque': self.quantidade_estoque,
            'dataCadastro': self.data_cadastro.isoformat(),
            'produtoMarcaModel': self.produto_marca_model.serialize() if self.produto_marca_model else None,
            'tributIcmsCustomCabModel': self.tribut_icms_custom_cab_model.serialize() if self.tribut_icms_custom_cab_model else None,
            'tributGrupoTributarioModel': self.tribut_grupo_tributario_model.serialize() if self.tribut_grupo_tributario_model else None,
            'produtoUnidadeModel': self.produto_unidade_model.serialize() if self.produto_unidade_model else None,
            'produtoSubgrupoModel': self.produto_subgrupo_model.serialize() if self.produto_subgrupo_model else None,
        }

File: frontend/compras/lib/app/data/model/produto_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:compras/app/infra/infra_imports.dart';
import 'package:compras/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class ProdutoModel {
	int? id;
	int? idProdutoSubgrupo;
	int? idProdutoMarca;
	int? idProdutoUnidade;
	int? idTributIcmsCustomCab;
	int? idTributGrupoTributario;
	String? nome;
	String? descricao;
	String? gtin;
	String? codigoInterno;
	double? valorCompra;
	double? valorVenda;
	String? codigoNcm;
	double? estoqueMinimo;
	double? estoqueMaximo;
	double? quantidadeEstoque;
	DateTime? dataCadastro;
	ProdutoMarcaModel? produtoMarcaModel;
	TributIcmsCustomCabModel? tributIcmsCustomCabModel;
	TributGrupoTributarioModel? tributGrupoTributarioModel;
	ProdutoUnidadeModel? produtoUnidadeModel;
	ProdutoSubgrupoModel? produtoSubgrupoModel;

	ProdutoModel({
		this.id,
		this.idProdutoSubgrupo,
		this.idProdutoMarca,
		this.idProdutoUnidade,
		this.idTributIcmsCustomCab,
		this.idTributGrupoTributario,
		this.nome,
		this.descricao,
		this.gtin,
		this.codigoInterno,
		this.valorCompra,
		this.valorVenda,
		this.codigoNcm,
		this.estoqueMinimo,
		this.estoqueMaximo,
		this.quantidadeEstoque,
		this.dataCadastro,
		this.produtoMarcaModel,
		this.tributIcmsCustomCabModel,
		this.tributGrupoTributarioModel,
		this.produtoUnidadeModel,
		this.produtoSubgrupoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'nome',
		'descricao',
		'gtin',
		'codigo_interno',
		'valor_compra',
		'valor_venda',
		'codigo_ncm',
		'estoque_minimo',
		'estoque_maximo',
		'quantidade_estoque',
		'data_cadastro',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Nome',
		'Descricao',
		'Gtin',
		'Codigo Interno',
		'Valor Compra',
		'Valor Venda',
		'Codigo Ncm',
		'Estoque Minimo',
		'Estoque Maximo',
		'Quantidade Estoque',
		'Data Cadastro',
	];

	ProdutoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idProdutoSubgrupo = jsonData['idProdutoSubgrupo'];
		idProdutoMarca = jsonData['idProdutoMarca'];
		idProdutoUnidade = jsonData['idProdutoUnidade'];
		idTributIcmsCustomCab = jsonData['idTributIcmsCustomCab'];
		idTributGrupoTributario = jsonData['idTributGrupoTributario'];
		nome = jsonData['nome'];
		descricao = jsonData['descricao'];
		gtin = jsonData['gtin'];
		codigoInterno = jsonData['codigoInterno'];
		valorCompra = jsonData['valorCompra']?.toDouble();
		valorVenda = jsonData['valorVenda']?.toDouble();
		codigoNcm = jsonData['codigoNcm'];
		estoqueMinimo = jsonData['estoqueMinimo']?.toDouble();
		estoqueMaximo = jsonData['estoqueMaximo']?.toDouble();
		quantidadeEstoque = jsonData['quantidadeEstoque']?.toDouble();
		dataCadastro = jsonData['dataCadastro'] != null ? DateTime.tryParse(jsonData['dataCadastro']) : null;
		produtoMarcaModel = jsonData['produtoMarcaModel'] == null ? ProdutoMarcaModel() : ProdutoMarcaModel.fromJson(jsonData['produtoMarcaModel']);
		tributIcmsCustomCabModel = jsonData['tributIcmsCustomCabModel'] == null ? TributIcmsCustomCabModel() : TributIcmsCustomCabModel.fromJson(jsonData['tributIcmsCustomCabModel']);
		tributGrupoTributarioModel = jsonData['tributGrupoTributarioModel'] == null ? TributGrupoTributarioModel() : TributGrupoTributarioModel.fromJson(jsonData['tributGrupoTributarioModel']);
		produtoUnidadeModel = jsonData['produtoUnidadeModel'] == null ? ProdutoUnidadeModel() : ProdutoUnidadeModel.fromJson(jsonData['produtoUnidadeModel']);
		produtoSubgrupoModel = jsonData['produtoSubgrupoModel'] == null ? ProdutoSubgrupoModel() : ProdutoSubgrupoModel.fromJson(jsonData['produtoSubgrupoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idProdutoSubgrupo'] = idProdutoSubgrupo != 0 ? idProdutoSubgrupo : null;
		jsonData['idProdutoMarca'] = idProdutoMarca != 0 ? idProdutoMarca : null;
		jsonData['idProdutoUnidade'] = idProdutoUnidade != 0 ? idProdutoUnidade : null;
		jsonData['idTributIcmsCustomCab'] = idTributIcmsCustomCab != 0 ? idTributIcmsCustomCab : null;
		jsonData['idTributGrupoTributario'] = idTributGrupoTributario != 0 ? idTributGrupoTributario : null;
		jsonData['nome'] = nome;
		jsonData['descricao'] = descricao;
		jsonData['gtin'] = gtin;
		jsonData['codigoInterno'] = codigoInterno;
		jsonData['valorCompra'] = valorCompra;
		jsonData['valorVenda'] = valorVenda;
		jsonData['codigoNcm'] = codigoNcm;
		jsonData['estoqueMinimo'] = estoqueMinimo;
		jsonData['estoqueMaximo'] = estoqueMaximo;
		jsonData['quantidadeEstoque'] = quantidadeEstoque;
		jsonData['dataCadastro'] = dataCadastro != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataCadastro!) : null;
		jsonData['produtoMarcaModel'] = produtoMarcaModel?.toJson;
		jsonData['tributIcmsCustomCabModel'] = tributIcmsCustomCabModel?.toJson;
		jsonData['tributGrupoTributarioModel'] = tributGrupoTributarioModel?.toJson;
		jsonData['produtoUnidadeModel'] = produtoUnidadeModel?.toJson;
		jsonData['produtoSubgrupoModel'] = produtoSubgrupoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idProdutoSubgrupo = plutoRow.cells['idProdutoSubgrupo']?.value;
		idProdutoMarca = plutoRow.cells['idProdutoMarca']?.value;
		idProdutoUnidade = plutoRow.cells['idProdutoUnidade']?.value;
		idTributIcmsCustomCab = plutoRow.cells['idTributIcmsCustomCab']?.value;
		idTributGrupoTributario = plutoRow.cells['idTributGrupoTributario']?.value;
		nome = plutoRow.cells['nome']?.value;
		descricao = plutoRow.cells['descricao']?.value;
		gtin = plutoRow.cells['gtin']?.value;
		codigoInterno = plutoRow.cells['codigoInterno']?.value;
		valorCompra = plutoRow.cells['valorCompra']?.value?.toDouble();
		valorVenda = plutoRow.cells['valorVenda']?.value?.toDouble();
		codigoNcm = plutoRow.cells['codigoNcm']?.value;
		estoqueMinimo = plutoRow.cells['estoqueMinimo']?.value?.toDouble();
		estoqueMaximo = plutoRow.cells['estoqueMaximo']?.value?.toDouble();
		quantidadeEstoque = plutoRow.cells['quantidadeEstoque']?.value?.toDouble();
		dataCadastro = Util.stringToDate(plutoRow.cells['dataCadastro']?.value);
		produtoMarcaModel = ProdutoMarcaModel();
		produtoMarcaModel?.nome = plutoRow.cells['produtoMarcaModel']?.value;
		tributIcmsCustomCabModel = TributIcmsCustomCabModel();
		tributIcmsCustomCabModel?.descricao = plutoRow.cells['tributIcmsCustomCabModel']?.value;
		tributGrupoTributarioModel = TributGrupoTributarioModel();
		tributGrupoTributarioModel?.descricao = plutoRow.cells['tributGrupoTributarioModel']?.value;
		produtoUnidadeModel = ProdutoUnidadeModel();
		produtoUnidadeModel?.sigla = plutoRow.cells['produtoUnidadeModel']?.value;
		produtoSubgrupoModel = ProdutoSubgrupoModel();
		produtoSubgrupoModel?.nome = plutoRow.cells['produtoSubgrupoModel']?.value;
	}	

	ProdutoModel clone() {
		return ProdutoModel(
			id: id,
			idProdutoSubgrupo: idProdutoSubgrupo,
			idProdutoMarca: idProdutoMarca,
			idProdutoUnidade: idProdutoUnidade,
			idTributIcmsCustomCab: idTributIcmsCustomCab,
			idTributGrupoTributario: idTributGrupoTributario,
			nome: nome,
			descricao: descricao,
			gtin: gtin,
			codigoInterno: codigoInterno,
			valorCompra: valorCompra,
			valorVenda: valorVenda,
			codigoNcm: codigoNcm,
			estoqueMinimo: estoqueMinimo,
			estoqueMaximo: estoqueMaximo,
			quantidadeEstoque: quantidadeEstoque,
			dataCadastro: dataCadastro,
		);			
	}

	
}